exporter/exporterhelper/internal/queue_capacity.go

```diff
@@ -0,0 +1,112 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+import (
+	"sync/atomic"
+)
+
+type itemsCounter interface {
+	ItemsCount() int
+}
+
+// QueueCapacityLimiter is an interface to control the capacity of the queue.
+type QueueCapacityLimiter[T any] interface {
+	// Capacity is the maximum capacity of the queue.
+	Capacity() int
+	// Size returns the current size of the queue.
+	Size() int
+	// claim tries to claim capacity for the given element. If the capacity is not available, it returns false.
+	claim(T) bool
+	// release releases capacity for the given queue element.
+	release(T)
+	// sizeOf returns the size of the given element.
+	sizeOf(T) uint64
+}
+
+type baseCapacityLimiter[T any] struct {
+	used *atomic.Uint64
+	cap  uint64
+}
+
+func newBaseCapacityLimiter[T any](capacity int) baseCapacityLimiter[T] {
+	return baseCapacityLimiter[T]{
+		used: &atomic.Uint64{},
+		cap:  uint64(capacity),
+	}
+}
+
+func (bcl *baseCapacityLimiter[T]) Capacity() int {
+	return int(bcl.cap)
+}
+
+func (bcl *baseCapacityLimiter[T]) Size() int {
+	return int(bcl.used.Load())
+}
+
+//nolint:unused
+func (bcl *baseCapacityLimiter[T]) claim(claimed uint64) bool {
+	if bcl.used.Load()+claimed > bcl.cap {
+		return false
+	}
+	bcl.used.Add(claimed)
+	return true
```

Comment:
This can overflow. To correctly do this, use this code:
```suggestion
	if bcl.used.Add(claimed) > bcl.cap {
		bcl.used.Add(-claimed)
		return false
	}
	return true
```

Reply:
Thanks. Updated